tests/integration/helm/common_test.go

```diff
@@ -112,3 +154,104 @@ func getBase64EncodedStringOfFile(filePath string) string {
+			kubeCtx = kc
+		} else {
+			fmt.Printf("The string '%s' does not contain the required context slug. Defaulting to minikube.\n", kc)
+		}
+	}
+	return kubeCtx
+}
+
+func renderTemplate(templateString string, data interface{}) (string, error) {
+	tmpl, err := template.New("resource").Parse(templateString)
+	if err != nil {
+		return "", err
+	}
+
+	var buffer bytes.Buffer
+	if err := tmpl.Execute(&buffer, data); err != nil {
+		return "", err
+	}
+
+	return buffer.String(), nil
+}
+
+func pvToYaml(pv PersistentVolume) (string, error) {
+	return renderTemplate(persistentVolumeYamlTpl, pv)
+}
+
+func pvcToYaml(pvc PersistentVolumeClaim) (string, error) {
+	return renderTemplate(persistentVolumeClaimYamlTpl, pvc)
+}
+
+func generateRandomString(length int) string {
+	const charset = "abcdefghijklmnopqrstuvwxyz"
+
+	// Seed the random number generator
+	rand.Seed(time.Now().UnixNano())
+
+	// Create a byte slice to store the random characters
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = charset[rand.Intn(len(charset))]
+	}
+	return string(b)
+}
+
+func enforceReady(subT *testing.T, kubectlOptions *k8s.KubectlOptions, vals []serviceValidations) {
+	// Pods might have to connect to each other. So, we should
+	// wait for all pods to be ready before doing any log checks.
+	waitTime := 5 * time.Second
+	retries := 96
+	for _, expected := range vals {
+		deployment := k8s.GetDeployment(subT, kubectlOptions, expected.name)
+		// when pulling images for the first time, it may take longer than 90s
+		// 360 seconds of retries. Pods typically ready in ~51 seconds if the image is already pulled.
+		k8s.WaitUntilDeploymentAvailable(subT, kubectlOptions, deployment.Name, retries, waitTime)
+
+		// Validate that k8s service is ready (pods are started and in service)
+		k8s.WaitUntilServiceAvailable(subT, kubectlOptions, expected.name, 10, 1*time.Second)
+	}
+}
+
+func checkPodLogsWithRetries(subT *testing.T, kubectlOptions *k8s.KubectlOptions, pods []corev1.Pod, tc string, svc string, expected string) {
+	// The pods report they're ready before the final log that we sometimes
+	// test. The root issue is that pods report ready before they truly are.
+	// Once that is fixed, this test becomes redundant and isn't required.
+	maxRetries := 3
+	retryDelay := 2 * time.Second
+
+	for _, pod := range pods {
+		var log string
+
+		for i := 0; i < maxRetries; i++ {
+			log = get_logs(subT, kubectlOptions, &pod, "")
+			if contains(log, expected) {
+				// Log entry found, proceed to next pod
+				break
+			}
+
+			// Log entry not found, wait before retrying
+			if i < maxRetries-1 {
+				time.Sleep(retryDelay)
+			}
+		}
+
+		// Final assertion
+		subT.Run(fmt.Sprintf("%s - %s", tc, svc), func(t *testing.T) {
+			if !contains(log, expected) {
+				t.Errorf("[ERROR]: %s - %s - log should contain matching entry:\n\t%s", tc, svc, expected)
+			}
+		})
+	}
+}
+
+func contains(s, substr string) bool {
+	return strings.Contains(s, substr)
+}
```

Comment:
What is the value of this function? saving typing a few extra characters?

Reply:
Removed extraneous fn